cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java

```diff
@@ -75,20 +82,23 @@ public void process(ApplicationWithPrograms input) throws Exception {
         continue;
       }
       ProgramType programType = ProgramType.valueOfSchedulableType(newScheduleSpec.getProgram().getProgramType());
+      ProgramId programId = appId.program(programType, newScheduleSpec.getProgram().getProgramName());
 
       // if the schedule differ in schedule type then we have deleted the existing one earlier in DeleteScheduleStage.
       // create it with the new type and spec here. See CDAP-8918 for details.
       if (ScheduleType.fromSchedule(newScheduleSpec.getSchedule()) !=
         ScheduleType.fromSchedule(oldScheduleSpec.getSchedule())) {
         LOG.debug("Redeploying schedule {} with specification {} which existed earlier with specification {}",
                   entry.getKey(), newScheduleSpec, oldScheduleSpec);
-        createSchedule(appId.program(programType, newScheduleSpec.getProgram().getProgramName()), newScheduleSpec);
+        createSchedule(programId, newScheduleSpec);
         continue;
       }
 
-      scheduler.updateSchedule(appId.program(programType, newScheduleSpec.getProgram().getProgramName()),
+      scheduler.updateSchedule(programId,
                                newScheduleSpec.getProgram().getProgramType(),
                                newScheduleSpec.getSchedule());
+      programScheduler.deleteSchedule(input.getApplicationId().schedule(oldScheduleSpec.getSchedule().getName()));
```

Comment:
perhaps we should have a updateSchedule() in Scheduler, that internally deletes and adds. That way, we don't hardcode here, how update is done.